### PR TITLE
[zos_find] Support GDG and special characters in data set names for zos_find

### LINKED ISSUE #1374: [Enabler] [zos_find] GDG/GDS and special character support
### Enabler description

```python
MVSDataSet(
            name = params.get("name"),
            record_format = params.get("record_format", None),
            volumes = params.get("volumes", None),
            data_set_type = params.get("type", None),
            block_size = params.get("block_size", None),
            record_length = params.get("record_length", None),
            space_primary = params.get("space_primary", None),
            space_secondary = params.get("space_secondary", None),
            space_type = params.get("space_type", None),
            directory_blocks = params.get("directory_blocks", None),
            key_length = params.get("key_length", None),
            key_offset = params.get("key_offset", None),
            sms_storage_class = params.get("sms_storage_class", None),
            sms_data_class = params.get("sms_data_class", None),
            sms_management_class = params.get("sms_management_class", None),
        )
```

GDG/GDS Support
-------------------

Option `patterns` and `pds_patterns` accept a list of string either as a pattern or a list of data set names, so user can provide a list of data set names that contain some GDS relative data set names, in that case, the module needs to resolve the relative name into absolute name before packing to correctly archive the given data set name.

```yaml
- name: Exclude data sets that have a low level qualifier 'TEST'
  zos_find:
    patterns: ZOAUSER.GDG(-1)
    contains: "hello"

```

Internally, in the case of a data set name being a GDS relative name we should resolve those names into absolute GDS names when the generation number is 0 or negative.

Following functions will be implemented in #1386 
```python
# Sample of soon to be implemented gdg function
if DataSet.is_gds_relative_name(name):
    src = DataSet.resolve_gds_absolute_name(src)
```

Symbols support
-----------------
 
Changes done to the BetterArgParser should be adopted in the module to not break and use the correct escaped name in case of the data set name having a symbol in it.

Add symbol support to the dest_data_set name option.

data set name with symbol

`ZOAUSER.TE$T`

That way, the modules will have to retrieve the new escaped named and the original user provided name inside the dictionary like:

```python
name = DataSet.escape_data_set_name(name)
```

Escaping special characters could be done via a new function inside the BetterArgParser.
These are some of the special characters:
- \#
- @
- $
- \-

For this task completion we need:
- [ ] Do the development work
- [ ] Update the docs to talk about GDG and GDS handling in the module.
- [ ] Update the docs with at least 1 example on how to use with GDSs.
- [ ] Create a new test case for GDSs.
- [ ] Create a new test case for symbols.

### Ansible module

zos_find

### LINKED ISSUE #1524: [Enhancement] allow zos_find to search and filter GDGs by attributes
### Is there an existing issue for this?

- [X] There are no existing issues.

### Enhancement or feature description

We want to expand zos_find to allow GDG-related attributes to be used as filter of a find.

Initial discussion with Fernando during #1374 review (#1518)

[fernandofloresg] /1518#discussion_r1612307235
Hey Rich, I don't know if this was added into the development task but I think we also should be able to filter the GDGs by its attributes, I mean, return GDGs with purge = True or return where extended = True. Is that something that can be added in this same task ?

### Ansible module

zos_find

### PR DESCRIPTION
**Pull request by richp405**
##### SUMMARY
Add support and test for GDG data set convention, and validate special-character use in ds names.

Fixes #1374
Fixes #1524 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
zos_find

### CODE REVIEW

**Comment by richp405** [comment]
image

**Comment by richp405** [comment]
I had to re-push, because dev changed while I was testing.

**Review by fernandofloresg — CHANGES_REQUESTED** [review]

File: tests/functional/modules/test_zos_find_func.py
Comment (fernandofloresg) [inline tests/functional/modules/test_zos_find_func.py:1]: Hey Rich, I don't know if this was added into the development task but I think we also should be able to filter the GDGs by its attributes, I mean, return GDGs with purge = True or return where extended = True. Is that something that can be added in this same task ?

Comment (richp405) [reply tests/functional/modules/test_zos_find_func.py:1]: I'll check on Friday for this... not sure, offhand, but we can touch up the interface.

File: tests/functional/modules/test_zos_find_func.py
Comment (richp405) [reply tests/functional/modules/test_zos_find_func.py:36]: I'll read through that... the reasons it was hardcoded before is that the testing was checking for the files with asterisks (like test.find.pds.*).. I think I can work up a compromise (like test.find.pds.randomname?)

File: tests/functional/modules/test_zos_find_func.py
Comment (richp405) [reply tests/functional/modules/test_zos_find_func.py:1]: It's a good thought, but I think we'll do it in a separate ticket, because changing the interface needs to be separate from adding options.  We can still fit this into 1.11-b1

File: tests/functional/modules/test_zos_find_func.py
Comment (richp405) [reply tests/functional/modules/test_zos_find_func.py:42]: I'm working through the change to run the tests against variable names.  good idea.  It will be ready Tuesday.

File: tests/functional/modules/test_zos_find_func.py
Comment (fernandofloresg) [reply tests/functional/modules/test_zos_find_func.py:1]: Sure, I noticed is not mentioned in the base issue, my bad.

**Review by fernandofloresg — CHANGES_REQUESTED** [review]
I code wise test is good. The only thing I would ask to add that is in the issue description is to modify the zos_apf docs to include the fact that GDGs are now supported.

File: tests/functional/modules/test_zos_find_func.py
Comment (richp405) [reply tests/functional/modules/test_zos_find_func.py:1]: Added #1524 as new enhancement to split up the work.

File: tests/functional/modules/test_zos_find_func.py
Comment (richp405) [reply tests/functional/modules/test_zos_find_func.py:42]: Make that Friday of this week.  I'll re-issue invites then.

File: tests/functional/modules/test_zos_find_func.py
Comment (fernandofloresg) [reply tests/functional/modules/test_zos_find_func.py:1]: Great thanks Rich!

File: tests/functional/modules/test_zos_find_func.py
Comment (fernandofloresg) [reply tests/functional/modules/test_zos_find_func.py:42]: Sure, I don't expect you to change all tests just this one to not introduce new hardcoded content.

**Review by fernandofloresg — APPROVED** [review]
I did some changes myself, hence moving from changes requested to approved. Will merge once we have 3 approvals

**Comment by fernandofloresg** [comment]
Screenshot 2024-07-14 at 10 32 15 PM

**Review by rexemin — APPROVED** [review]

**Pull request merged**
